resources/prosody-plugins/mod_reservations.lua

```diff
@@ -0,0 +1,592 @@
+        self:handle_conference_already_exist(response_body);
+    elseif response_code == nil then  -- warrants a retry, but this should be done automatically by the http call method.
+        self:set_status_failed(500, 'Could not contact reservation server');
+    else
+        self:set_status_failed(response_code, self:parse_error_message_from_response(response_body));
+    end
+end
+
+function RoomReservation:handler_conference_data_returned_from_api(response_body)
+    local data = self:parse_conference_response(response_body);
+    if not data then  -- invalid response from API
+        module:log("error", "API returned success code but invalid payload");
+        self:set_status_failed(500, 'Invalid response from reservation server');
+    else
+        self:set_status_success(data.start_time, data.duration, data.mail_owner, data.id)
+    end
+end
+
+function RoomReservation:handle_conference_already_exist(response_body)
+    local data = json.decode(response_body);
+    if data == nil or data.conflict_id == nil then
+        -- yes, in the case of 409, API expected to return "id" as "conflict_id".
+        self:set_status_failed(409, 'Invalid response from reservation server');
+    else
+        local url = api_prefix..'/conference/'..data.conflict_id;
+        local http_options = {
+            method = 'GET';
+            headers = http_headers;
+        }
+
+        async_http_request(url, http_options, function(response_body, response_code)
+            if response_code == 200 then
+                self:handler_conference_data_returned_from_api(response_body);
+            else
+                self:set_status_failed(response_code, self:parse_error_message_from_response(response_body));
+            end
+        end, function ()
+            self:on_api_call_timeout();
+        end);
+    end
+end
+
+--- End RoomReservation
+
+--- Store reservations lookups that are still pending or with room still active
+local reservations = {}
+
+local function get_or_create_reservations(room_jid, creator_jid)
+    if reservations[room_jid] == nil then
+        module:log("debug", "Creating new reservation data for %s", room_jid);
+        reservations[room_jid] = newRoomReservation(room_jid, creator_jid);
+    end
+
+    return reservations[room_jid];
+end
+
+local function evict_expired_reservations()
+    local expired = {}
+
+    -- first, gather jids of expired rooms. So we don't remove from table while iterating.
+    for room_jid, res in pairs(reservations) do
+        if res:is_expired() then
+            table.insert(expired, room_jid);
+        end
+    end
+
+    local room;
+    for _, room_jid in ipairs(expired) do
+        room = get_room_from_jid(room_jid);
+        if room then
+            -- Close room if still active (reservation duration exceeded)
+            module:log("info", "Room exceeded reservation duration. Terminating %s", room_jid);
+            room:clear();
+            room:destroy();
```

Comment:
Hey, @shawnchin do you remember why you added room:clear here?

Dropping the clear and just changing `room:destroy(nil, "Scheduled conference duration exceeded.");` fixes the destroy problem.

Reply:
Don't recall any good reason. It's likely I picked it up from somewhere in the community forum.

Thanks for fixing :)